src/ornament.js

```diff
@@ -27,27 +27,100 @@ export class Ornament extends Modifier {
+          left_shift += reportedWidth;
+        } else if (ornament.render_options.xOffset > 0) {
+          right_shift += reportedWidth;
+        }
 
-      if (ornament.getPosition() === Modifier.Position.ABOVE) {
-        ornament.setTextLine(state.top_text_line);
-        state.top_text_line += increment;
-      } else {
-        ornament.setTextLine(state.text_line);
-        state.text_line += increment;
+        // articulations above/below the line can be stacked.
+        if (Ornament.ornamentArticulation.indexOf(ornament.type) >= 0) {
+          // Unfortunately we con't know the stem direction.  So we base it
+          // on the line number, but also allow it to be overridden.
+          if (ornament.note.getLineNumber() >= 3 || ornament.getPosition() === Modifier.Position.ABOVE) {
+            state.top_text_line += increment;
+            ornament.render_options.yOffset += yOffset;
+            yOffset -= ornament.glyph.bbox.h;
+          } else {
+            state.text_line += increment;
+            ornament.render_options.yOffset += yOffset;
+            yOffset += ornament.glyph.bbox.h;
+          }
+        }
+      } else { // logic for classical ornament formatting.
+        width = Math.max(ornament.getWidth(), width);
+        if (ornament.getPosition() === Modifier.Position.ABOVE) {
+          ornament.setTextLine(state.top_text_line);
+          state.top_text_line += increment;
+        } else {
+          ornament.setTextLine(state.text_line);
+          state.text_line += increment;
+        }
       }
     }
-
-    state.left_shift += width / 2;
-    state.right_shift += width / 2;
+    // Note: 'legit' ornaments don't consider other modifiers when calculating their
+    // X position, but jazz ornaments sometimes need to.
+    state.left_shift = left_shift + (width / 2);
+    state.right_shift = right_shift + (width / 2);
     return true;
   }
 
+  // ### ornamentNoteTransition
+  // means the jazz ornament represents an effect from one note to another,
+  // these are generally on the top of the staff.
+  static get ornamentNoteTransition() {
+    return ['flip', 'jazzTurn', 'smear'];
+  }
+
+  // ### ornamentAttack
+  // Indicates something that happens in the attach, placed before the note and
+  // any accidentals
+  static get ornamentAttack() {
+    return ['scoop'];
+  }
+
+  // ### ornamentRelease
+  // An ornament that happens on the release of the note, generally placed after the
+  // note and overlapping the next beat/measure..
+  static get ornamentRelease() {
+    return [
+      'doit', 'fall', 'fallLong', 'doitLong', 'jazzTurn', 'smear', 'flip'
+    ];
+  }
+
+  // ### ornamentArticulation
+  // goes above/below the note based on space availablity
+  static get ornamentArticulation() {
+    return ['bend', 'plungerClosed', 'plungerOpen'];
+  }
+
+  static get glyphMetrics() {
+    return Vex.Flow.DEFAULT_FONT_STACK[0].metrics.glyphs.jazzOrnaments;
+  }
+
+  get metrics() {
```

Comment:
We don't use (non-static) metrics as convention, can you replace this with `getMetrics()` (this keeps the codebase consistent.)